### DISCUSSION TITLE
Dynamically modify Forms using AJAX without jQuery

### DISCUSSION THREAD

**Discussion by soleneggd** [symfony/symfony#47800] [Q&A] [answered]
Hello, I'm trying to dynamically modify forms using form events as describe on this page : [How to Dynamically Modify Forms Using Form Events](https://symfony.com/doc/current/form/dynamic_form_modification.html)

Everything works fine but for optimisation reasons, I try not to use jQuery anymore, which is used in the last part of the doc.
Here is my script using jQuery which is almost the same as in the doc (this script works fine and I get the new value in my field) : 
```js
var $field_1 = $("#field_1_id");
$field_1.change(function() {
    var $form = $(this).closest("form");
    var data = {};
    data[$field_1.attr("name")] = $field_1.val();
    $.ajax({
        url : $form.attr("action"),
        type: $form.attr("method"),
        data : data,
        success: function(html) {
            $("#field_2_id").replaceWith(
                $(html).find("#field_2_id")
            );
        }
    });
});
```

Here is my equivalent in Vanilla Javascript : 
```js
document.getElementById("field_1_id").addEventListener("change", (event) => {
    let field_1 = event.target;
    let form = field_1.closest("form");
    const xhttp = new XMLHttpRequest();
    xhttp.open(form.getAttribute("method"), form.getAttribute("action") ?? window.location.href, true);
    xhttp.setRequestHeader("Content-Type", "application/json");
    xhttp.onreadystatechange = function() {
        if (this.readyState == 4 && this.status == 200) {
            let html = document.createElement("div");
            html.innerHTML = this.responseText;
            let old_field = document.getElementById("field_2_id");
            let new_field= html.querySelector("#field_2_id");
            old_field.parentNode.replaceChild(new_field, old_field);
        }
    }
    let data = {};
    data[field_1.getAttribute("name")] = field_1.value;
    xhttp.send(JSON.stringify(data));
});
```
But with this script, I don't get anything in my field. I get the form in the response but with no data.
It seems that my EventListener (`FormEvents::PRE_SET_DATA`) does not set the new value of my field 2 because it don't get the value of my field 1 (as if my script did not send any data).
I think the problem is in my script and not in my Controller because it works fine with jQuery.

I specify that I use `form.getAttribute("action") ?? window.location.href` as URL because I don't have any action attribute in my form for the moment, and the url param of the jQuery ajax function automatically use the current page url if the value is undefined.

Does anyone have any idea what I'm doing wrong please ?

**Comment by stof** [comment, marked as answer]
The issue is that your code sends a JSON-encoded body, not a form-encoded body.

Also note that the handling of the GET and POST method needs to be different. For a GET form, the form data goes in the URL, not in the body (jQuery's AJAX method supports using the `data` option with the GET method and will handle updating the URL, but native APIs don't have such magic).

Here is a vanilla JS equivalent (note that I'm using modern JS features because it is much simpler that using XMLHttpRequest or building the URL or the form-encoded body manually with all the proper escaping):

```js
document.getElementById("field_1_id").addEventListener("change", async (event) => {
    const field_1 = event.target;
    const form = field_1.closest("form");
    let response
    if (form.method === 'get') {
        const url = new URL(form.action)
        url.searchParams.set(field_1.name, field_1.value)
        response = await fetch(url)
    } else {
        const data = new FormData()
        data.set(field_1.name, field_1.value) // Or use `data = new FormData(form)` to submit the data for the whole form
        response = await fetch(form.action, { method: form.method, body: data })
    }

    if (response.status === 200) {
            const html = document.createElement("div")
            html.innerHTML = await response.text()
            const old_field = document.getElementById("field_2_id");
            const new_field= html.querySelector("#field_2_id");
            old_field.parentNode.replaceChild(new_field, old_field);
    }
});
```

**Reply by soleneggd** [reply]
It works perfectly, thank you for your quick reply !! 😄
I didn't think about forms that uses GET method, thank you for reminding me.
